bindings/rust/s2n-tls/src/raw/connection.rs

```diff
@@ -18,17 +18,88 @@ use libc::c_void;
 use s2n_tls_sys::*;
 use std::{ffi::CStr, mem};
 
-pub use s2n_tls_sys::s2n_mode;
+use s2n_tls_sys::s2n_mode;
+
+macro_rules! static_const_str {
+    ($c_chars:expr) => {
+        unsafe { CStr::from_ptr($c_chars) }
+            .to_str()
+            .map_err(|_| Error::InvalidInput)
+    };
+}
+
+#[derive(Debug, PartialEq)]
+pub enum CallbackResult {
+    Success,
+    Failure,
+}
+
+impl From<CallbackResult> for s2n_status_code::Type {
+    fn from(input: CallbackResult) -> s2n_status_code::Type {
+        match input {
+            CallbackResult::Success => s2n_status_code::SUCCESS,
+            CallbackResult::Failure => s2n_status_code::FAILURE,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Mode {
+    Server,
+    Client,
+}
+
+impl From<Mode> for s2n_mode::Type {
+    fn from(input: Mode) -> s2n_mode::Type {
+        match input {
+            Mode::Server => s2n_mode::SERVER,
+            Mode::Client => s2n_mode::CLIENT,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Version {
+    Unknown,
+    SSLV2,
+    SSLV3,
+    TLS10,
+    TLS11,
+    TLS12,
+    TLS13,
+}
+
+impl From<s2n_tls_version::Type> for Version {
+    fn from(input: s2n_tls_version::Type) -> Self {
+        match input {
+            s2n_tls_version::SSLV2 => Self::SSLV2,
+            s2n_tls_version::SSLV3 => Self::SSLV3,
+            s2n_tls_version::TLS10 => Self::TLS10,
+            s2n_tls_version::TLS11 => Self::TLS11,
+            s2n_tls_version::TLS12 => Self::TLS12,
+            s2n_tls_version::TLS13 => Self::TLS13,
+            _ => Self::Unknown,
```

Comment:
do we expect `Unknown`? should this instead be a `panic!()`?

I think we also want to add `#[non_exhaustive]` to the enum to capture that new version can be added later.

Reply:
It definitely shouldn't just panic-- you can get the s2n-tls internal version of "unknown" from a server that hasn't performed a handshake yet.

I could implement it as try_from instead of from though. I went back and forth on that :\